TestResetTS is unstable

## Flaky Test

### Which jobs are failing
```
2023-11-03T02:17:32.7014932Z --- FAIL: TestAdminTestSuite (33.40s)
2023-11-03T02:17:32.7015364Z     --- FAIL: TestAdminTestSuite/TestResetTS (0.03s)
2023-11-03T02:17:32.7015779Z         api_check.go:29: 
2023-11-03T02:17:32.7016898Z             	Error Trace:	/home/runner/work/pd/pd/pkg/utils/testutil/api_check.go:29
2023-11-03T02:17:32.7019320Z             	            				/home/runner/work/pd/pd/pkg/utils/testutil/api_check.go:142
2023-11-03T02:17:32.7020790Z             	            				/home/runner/work/pd/pd/pkg/utils/testutil/api_check.go:105
2023-11-03T02:17:32.7022287Z             	            				/home/runner/work/pd/pd/server/api/admin_test.go:191
2023-11-03T02:17:32.7022608Z             	Error:      	Not equal: 
2023-11-03T02:17:32.7023173Z             	            	expected: 200
2023-11-03T02:17:32.7023568Z             	            	actual  : 403
2023-11-03T02:17:32.7024091Z             	Test:       	TestAdminTestSuite/TestResetTS
2023-11-03T02:17:32.7025089Z             	Messages:   	resp: "[PD:etcd:ErrEtcdTxnConflict]etcd transaction failed, conflicted and rolled back"
```

### CI link
https://github.com/tikv/pd/actions/runs/6740464608/job/18323580850

### Reason for failure (if possible)


### Anything else

```
2023-11-15T14:07:37.9609463Z --- FAIL: TestAdminTestSuite (15.55s)
2023-11-15T14:07:37.9610157Z     --- FAIL: TestAdminTestSuite/TestResetTS (0.11s)
2023-11-15T14:07:37.9610789Z         api_check.go:29: 
2023-11-15T14:07:37.9611577Z             	Error Trace:	/home/runner/work/pd/pd/pkg/utils/testutil/api_check.go:29
2023-11-15T14:07:37.9612824Z             	            				/home/runner/work/pd/pd/pkg/utils/testutil/api_check.go:157
2023-11-15T14:07:37.9614042Z             	            				/home/runner/work/pd/pd/pkg/utils/testutil/api_check.go:105
2023-11-15T14:07:37.9615039Z             	            				/home/runner/work/pd/pd/server/api/admin_test.go:216
2023-11-15T14:07:37.9615670Z             	Error:      	Not equal: 
2023-11-15T14:07:37.9616291Z             	            	expected: 403
2023-11-15T14:07:37.9616861Z             	            	actual  : 500
2023-11-15T14:07:37.9617441Z             	Test:       	TestAdminTestSuite/TestResetTS
2023-11-15T14:07:37.9618106Z             	Messages:   	resp: redirect to not leader
```

--- a/pkg/mcs/tso/server/apis/v1/api.go
+++ b/pkg/mcs/tso/server/apis/v1/api.go
@@ -15,6 +15,7 @@
 package apis
 
 import (
+	"fmt"
 	"net/http"
 	"strconv"
 	"sync"
@@ -150,6 +151,7 @@ type ResetTSParams struct {
 // @Failure  400  {string}  string  "The input is invalid."
 // @Failure  403  {string}  string  "Reset ts is forbidden."
 // @Failure  500  {string}  string  "TSO server failed to proceed the request."
+// @Failure  503  {string}  string  "It's a temporary failure, please retry."
 // @Router   /admin/reset-ts [post]
 // if force-use-larger=true:
 //
@@ -189,7 +191,8 @@ func ResetTS(c *gin.Context) {
 			// If the error is ErrEtcdTxnConflict, it means there is a temporary failure.
 			// Return 503 to let the client retry.
 			// Ref: https://datatracker.ietf.org/doc/html/rfc7231#section-6.6.4
-			c.String(http.StatusServiceUnavailable, err.Error())
+			c.String(http.StatusServiceUnavailable,
+				fmt.Sprintf("It's a temporary failure with error %s, please retry.", err.Error()))
 		} else {
 			c.String(http.StatusForbidden, err.Error())
 		}

--- a/pkg/mcs/tso/server/apis/v1/api.go
+++ b/pkg/mcs/tso/server/apis/v1/api.go
@@ -185,6 +185,8 @@ func ResetTS(c *gin.Context) {
 	if err = svr.ResetTS(ts, ignoreSmaller, skipUpperBoundCheck, 0); err != nil {
 		if err == errs.ErrServerNotStarted {
 			c.String(http.StatusInternalServerError, err.Error())
+		} else if err == errs.ErrEtcdTxnConflict {
+			c.String(http.StatusServiceUnavailable, err.Error())
 		} else {
 			c.String(http.StatusForbidden, err.Error())
 		}

--- a/pkg/mcs/tso/server/apis/v1/api.go
+++ b/pkg/mcs/tso/server/apis/v1/api.go
@@ -186,6 +186,9 @@ func ResetTS(c *gin.Context) {
 		if err == errs.ErrServerNotStarted {
 			c.String(http.StatusInternalServerError, err.Error())
 		} else if err == errs.ErrEtcdTxnConflict {
+			// If the error is ErrEtcdTxnConflict, it means there is a temporary failure.
+			// Return 503 to let the client retry.
+			// Ref: https://datatracker.ietf.org/doc/html/rfc7231#section-6.6.4
 			c.String(http.StatusServiceUnavailable, err.Error())
 		} else {
 			c.String(http.StatusForbidden, err.Error())

--- a/pkg/tso/admin.go
+++ b/pkg/tso/admin.go
@@ -97,6 +97,9 @@ func (h *AdminHandler) ResetTS(w http.ResponseWriter, r *http.Request) {
 		if err == errs.ErrServerNotStarted {
 			h.rd.JSON(w, http.StatusInternalServerError, err.Error())
 		} else if err == errs.ErrEtcdTxnConflict {
+			// If the error is ErrEtcdTxnConflict, it means there is a temporary failure.
+			// Return 503 to let the client retry.
+			// Ref: https://datatracker.ietf.org/doc/html/rfc7231#section-6.6.4
 			h.rd.JSON(w, http.StatusServiceUnavailable, err.Error())
 		} else {
 			h.rd.JSON(w, http.StatusForbidden, err.Error())

--- a/pkg/tso/admin.go
+++ b/pkg/tso/admin.go
@@ -15,6 +15,7 @@
 package tso
 
 import (
+	"fmt"
 	"net/http"
 	"strconv"
 
@@ -53,6 +54,7 @@ func NewAdminHandler(handler Handler, rd *render.Render) *AdminHandler {
 // @Failure  400  {string}  string  "The input is invalid."
 // @Failure  403  {string}  string  "Reset ts is forbidden."
 // @Failure  500  {string}  string  "TSO server failed to proceed the request."
+// @Failure  503  {string}  string  "It's a temporary failure, please retry."
 // @Router   /admin/reset-ts [post]
 // if force-use-larger=true:
 //
@@ -100,7 +102,8 @@ func (h *AdminHandler) ResetTS(w http.ResponseWriter, r *http.Request) {
 			// If the error is ErrEtcdTxnConflict, it means there is a temporary failure.
 			// Return 503 to let the client retry.
 			// Ref: https://datatracker.ietf.org/doc/html/rfc7231#section-6.6.4
-			h.rd.JSON(w, http.StatusServiceUnavailable, err.Error())
+			h.rd.JSON(w, http.StatusServiceUnavailable,
+				fmt.Sprintf("It's a temporary failure with error %s, please retry.", err.Error()))
 		} else {
 			h.rd.JSON(w, http.StatusForbidden, err.Error())
 		}

--- a/pkg/tso/admin.go
+++ b/pkg/tso/admin.go
@@ -96,6 +96,8 @@ func (h *AdminHandler) ResetTS(w http.ResponseWriter, r *http.Request) {
 	if err = handler.ResetTS(ts, ignoreSmaller, skipUpperBoundCheck, 0); err != nil {
 		if err == errs.ErrServerNotStarted {
 			h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+		} else if err == errs.ErrEtcdTxnConflict {
+			h.rd.JSON(w, http.StatusServiceUnavailable, err.Error())
 		} else {
 			h.rd.JSON(w, http.StatusForbidden, err.Error())
 		}

--- a/server/api/admin_test.go
+++ b/server/api/admin_test.go
@@ -18,6 +18,7 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"io"
 	"net/http"
 	"testing"
 	"time"
@@ -27,6 +28,7 @@ import (
 	"github.com/stretchr/testify/suite"
 	"github.com/tikv/pd/pkg/core"
 	"github.com/tikv/pd/pkg/replication"
+	"github.com/tikv/pd/pkg/utils/apiutil"
 	tu "github.com/tikv/pd/pkg/utils/testutil"
 	"github.com/tikv/pd/server"
 )
@@ -188,9 +190,24 @@ func (suite *adminTestSuite) TestResetTS() {
 	values, err := json.Marshal(args)
 	suite.NoError(err)
 	re := suite.Require()
-	err = tu.CheckPostJSON(testDialClient, url, values,
-		tu.StatusOK(re),
-		tu.StringEqual(re, "\"Reset ts successfully.\"\n"))
+	tu.Eventually(re, func() bool {
+		resp, err := apiutil.PostJSON(testDialClient, url, values)
+		re.NoError(err)
+		defer resp.Body.Close()
+		b, err := io.ReadAll(resp.Body)
+		re.NoError(err)
+		switch resp.StatusCode {
+		case http.StatusOK:
+			re.Contains(string(b), "Reset ts successfully.")
+			return true
+		case http.StatusForbidden:
+			re.Contains(string(b), "[PD:etcd:ErrEtcdTxnConflict]etcd transaction failed, conflicted and rolled back")
+			return false
+		default:
+			re.FailNow("unexpected status code %d", resp.StatusCode)
+			return false
+		}
+	})
 	suite.NoError(err)
 	t2 := makeTS(32 * time.Hour)
 	args["tso"] = fmt.Sprintf("%d", t2)

--- a/server/api/admin_test.go
+++ b/server/api/admin_test.go
@@ -203,6 +203,9 @@ func (suite *adminTestSuite) TestResetTS() {
 		case http.StatusForbidden:
 			re.Contains(string(b), "[PD:etcd:ErrEtcdTxnConflict]etcd transaction failed, conflicted and rolled back")
 			return false
+		case http.StatusInternalServerError:
+			re.Contains(string(b), "redirect to not leader")
+			return false
 		default:
 			re.FailNow("unexpected status code %d", resp.StatusCode)
 			return false

--- a/server/api/admin_test.go
+++ b/server/api/admin_test.go
@@ -200,11 +200,8 @@ func (suite *adminTestSuite) TestResetTS() {
 		case http.StatusOK:
 			re.Contains(string(b), "Reset ts successfully.")
 			return true
-		case http.StatusForbidden:
+		case http.StatusServiceUnavailable:
 			re.Contains(string(b), "[PD:etcd:ErrEtcdTxnConflict]etcd transaction failed, conflicted and rolled back")
-			return false
-		case http.StatusInternalServerError:
-			re.Contains(string(b), "redirect to not leader")
 			return false
 		default:
 			re.FailNow("unexpected status code %d", resp.StatusCode)

